hmy/downloader/shortrange.go

```diff
@@ -0,0 +1,448 @@
+	for i := 0; i != sh.config.Concurrency; i++ {
+		go func() {
+			defer wg.Done()
+			defer cancel() // it's ok to cancel context more than once
+
+			for {
+				if m.isDone() {
+					return
+				}
+				hashes, wl, err := m.getNextHashes()
+				if err != nil {
+					errLock.Lock()
+					gErr = err
+					errLock.Unlock()
+					return
+				}
+				if len(hashes) == 0 {
+					select {
+					case <-time.After(200 * time.Millisecond):
+						continue
+					case <-ctx.Done():
+						return
+					}
+				}
+				blocks, stid, err := sh.doGetBlocksByHashesRequest(ctx, hashes, wl)
+				if err != nil {
+					sh.logger.Err(err).Msg("getBlocksByHashes worker failed")
+					m.handleResultError(hashes, stid)
+				} else {
+					m.addResult(hashes, blocks, stid)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+
+	if gErr != nil {
+		return nil, gErr
+	}
+	select {
+	case <-sh.ctx.Done():
+		return nil, sh.ctx.Err()
+	default:
+	}
+
+	return m.getResults()
+}
+
+func (sh *srHelper) checkPrerequisites() error {
+	if sh.syncProtocol.NumStreams() < sh.config.Concurrency {
+		return errors.New("not enough streams")
+	}
+	return nil
+}
+
+func (sh *srHelper) prepareBlockHashNumbers(curNumber uint64) []uint64 {
+	res := make([]uint64, 0, numBlockHashesPerRequest)
+
+	for bn := curNumber + 1; bn <= curNumber+uint64(numBlockHashesPerRequest); bn++ {
+		res = append(res, bn)
+	}
+	return res
+}
+
+func (sh *srHelper) doGetBlockHashesRequest(bns []uint64) ([]common.Hash, sttypes.StreamID, error) {
+	ctx, cancel := context.WithTimeout(sh.ctx, 1*time.Second)
+	defer cancel()
+
+	hashes, stid, err := sh.syncProtocol.GetBlockHashes(ctx, bns)
+	if err != nil {
+		return nil, stid, err
+	}
+	if len(hashes) != len(bns) {
+		err := errors.New("unexpected get block hashes result delivered")
+		sh.logger.Warn().Err(err).Str("stream", string(stid)).Msg("failed to doGetBlockHashesRequest")
+		sh.syncProtocol.RemoveStream(stid)
+		return nil, stid, err
+	}
+	return hashes, stid, nil
+}
+
+func (sh *srHelper) doGetBlocksByHashesRequest(ctx context.Context, hashes []common.Hash, wl []sttypes.StreamID) ([]*types.Block, sttypes.StreamID, error) {
+	ctx, cancel := context.WithTimeout(sh.ctx, 10*time.Second)
+	defer cancel()
+
+	blocks, stid, err := sh.syncProtocol.GetBlocksByHashes(ctx, hashes,
+		syncProto.WithWhitelist(wl))
+	if err != nil {
+		return nil, stid, err
+	}
+	if err := checkGetBlockByHashesResult(blocks, hashes); err != nil {
+		sh.logger.Warn().Err(err).Str("stream", string(stid)).Msg("failed to getBlockByHashes")
+		sh.syncProtocol.RemoveStream(stid)
+		return nil, stid, err
+	}
+	return blocks, stid, nil
+}
+
+func (sh *srHelper) removeStreams(sts []sttypes.StreamID) {
+	for _, st := range sts {
+		sh.syncProtocol.RemoveStream(st)
+	}
+}
+
+func checkGetBlockByHashesResult(blocks []*types.Block, hashes []common.Hash) error {
+	if len(blocks) != len(hashes) {
+		return errors.New("unexpected number of getBlocksByHashes result")
+	}
+	for i, block := range blocks {
```

Comment:
assumed sorted list, how to guarantee that?

Reply:
Because the `hashChain` was assumed to be sorted. If the given hashChain is not sorted, remote node can be malicious and need to be killed.